### PR TITLE
Us002 view job specification

### PR DESCRIPTION
**Pull request by kamil-biegaj**

### CODE REVIEW

**Review by philipwilson5 — COMMENTED** [review]
Same issues as in us002

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,7 @@ target/
 .idea/jarRepositories.xml
 .idea/compiler.xml
 .idea/libraries/
+.idea/stat.log
```
Comment (shaunganley) [inline .gitignore]: you should .gitignore the entire .idea directory rather than the files

File: pom.xml [hunk 2]
```diff
@@ -87,6 +87,26 @@
...
+            <scope>test</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.mockito/mockito-junit-jupiter -->
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-junit-jupiter</artifactId>
+            <version>2.17.0</version>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (shaunganley) [inline pom.xml]: surely these already exist on main, can you make sure your branch is up to date with latest main?

File: src/java/com/example/api/perf/test/integrationTests/StudentControllerMockTests.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.Optional;
+
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.when;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class StudentControllerMockTests {
+
```
Comment (shaunganley) [inline src/java/com/example/api/perf/test/integrationTests/StudentControllerMockTests.java]: what is this file?

File: src/java/com/example/api/perf/test/integrationTests/StudentControllerMockmvcTests.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+public class StudentControllerMockmvcTests {
```
Comment (shaunganley) [inline src/java/com/example/api/perf/test/integrationTests/StudentControllerMockmvcTests.java]: what is this file?

File: src/java/com/example/api/perf/test/integrationTests/StudentControllerTests.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.util.Objects;
+
+import static org.junit.Assert.assertTrue;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class StudentControllerTests {
```
Comment (shaunganley) [inline src/java/com/example/api/perf/test/integrationTests/StudentControllerTests.java]: what is this file?

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 6]
```diff
@@ -35,4 +41,19 @@ public Response getAllJobRoles() {
...
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getJobSpecificationId(@PathParam("id") int id) {
+
+        try {
+            return Response.ok(jobRoleService.getSpecificationById(id)).build();
+        } catch (FailedToGetValidJobId e) {
+            System.err.println(e.getMessage());
+
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500).build();
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java]: there should be another exception scenario for the ID not existing in the DB which should return a 404

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 7]
```diff
@@ -27,4 +29,23 @@ public List<JobRole> getAllJobRoles() throws FailedToGetJobRolesException {
...
+        if (id < 0) {
+            throw new FailedToGetValidJobId();
+        }
+
+        try {
+            JobRoleSpec Spec = jobRoleDao.getSpecificationById(id, databaseConnector.getConnection());
+            if (Spec == null) {
+                throw new FailedToGetValidJobId();
+            }
+            return Spec;
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/JobRoleService.java]: you shouldn't use the same exception for JobRoleDoesNotExist as the SQLException below as these should return 2 different error codes from the API

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 8]
```diff
@@ -27,4 +29,23 @@ public List<JobRole> getAllJobRoles() throws FailedToGetJobRolesException {
...
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToGetValidJobId();
+        } catch (DatabaseConnectionException e) {
+            throw new RuntimeException(e);
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/JobRoleService.java]: don't throw generic exceptions, you can throw DatabaseConnectionException from this method and catch it in the controller

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                .get();
+        assertEquals(200, response.getStatus());
+        assertNotNull(response.getEntity());
+
+        // Request with invalid id
+        response = ClientBuilder.newClient()
+                .target(webServer + "/api/job_roles/invalidId")
+                .request(MediaType.APPLICATION_JSON)
+                .get();
+        assertEquals(404, response.getStatus());
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java]: firstly, these should be separate tests in better names test methods. secondly, this test won't pass as you never return 404 from your endpoint

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    public void testGetJobSpecificationId() {
+        // Request with valid id
+
+        Response response = ClientBuilder.newClient()
+                .target(webServer + "/api/job_roles/1")
+                .request(MediaType.APPLICATION_JSON)
+                .get();
+        assertEquals(200, response.getStatus());
+        assertNotNull(response.getEntity());
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java]: can you also assert that the ID of the spec returned is the same as the one you request in the URL

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+public class JobRoleServiceTest {
+    JobRoleDao jobRoleDao = Mockito.mock(JobRoleDao.class);
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+
+    JobRoleService jobRoleService = new JobRoleService(jobRoleDao, databaseConnector);
+
+    Connection conn;
+
+
+
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java]:
```suggestion

```

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        // Assuming you have a mock/stub implementation of JobRoleDao
+        Mockito.when(jobRoleDao.getSpecificationById(id, databaseConnector.getConnection())).thenReturn(expectedSpec);
+
+        // Act
+        JobRoleSpec actualSpec = jobRoleService.getSpecificationById(id); // Use jobRoleService instance here
+
+        // Assert
+        Assert.assertEquals(expectedSpec, actualSpec);
+    }
+}
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java]: you don't performance test in your unit tests, please remove

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        jobRoleService.getSpecificationById(id); // Use jobRoleService instance here
+    }
+
+    @Test(expected = FailedToGetValidJobId.class)
+    public void testGetSpecificationById_DatabaseException_ThrowsFailedToGetValidJobId() throws FailedToGetValidJobId, SQLException, DatabaseConnectionException {
+        // Arrange
+        int id = 5;
+        // Assuming you have a mock/stub implementation of JobRoleDao
+
+        Mockito.when(jobRoleDao.getSpecificationById(id, databaseConnector.getConnection())).thenThrow(new SQLException("Database connection error"));
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java]:
```suggestion
        Mockito.when(jobRoleDao.getSpecificationById(id, databaseConnector.getConnection())).thenThrow(SQLException.class));
```

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
+
+package com.example.api.perf.test.integrationTests;
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleIntegrationTest.java]: this shouldn't be in a perf test directory

File: src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
+package com.example.api.perf.test.integrationTests;
```
Comment (shaunganley) [inline src/test/java/com/example/api/perf/test/integrationTests/JobRoleServiceTest.java]: this shouldn't be in a perf test directory

**Review by shaunganley — CHANGES_REQUESTED** [review]
good start. missing controller unit tests and looks like there's some files copied across that shouldn't be in here

**Comment by kamil-biegaj** [comment]
@ccurran01

**Review by AlexMcC02 — APPROVED** [review]
Looks good!

**Review by ccurran01 — APPROVED** [review]

**Pull request merged**